File: bootstrap_src/com/zbyd/bootstrap/JdwpInject.java
package com.zbyd.bootstrap;

import java.nio.ByteBuffer;

/**
 * On-device JDWP in-process dex injector (Java port of jdwp_inject.py) over an ADB jdwp:<pid> stream.
 * Loads HudPrivAgent into the debuggable platform_app com.byd.cameraautostudy and calls start(), so the
 * agent runs with the signature BYDAUTO_*_SET perms after a reboot without a PC.
 *
 * Mirrors the validated PC injector:
 *  - breakpoint on Handler.dispatchMessage; the caller pokes with a config change (cmd uimode) so the
 *    MAIN thread dispatches a message -> a thread suspended at a Java safe point.
 *  - ReferenceType.Methods (2,5) reply has NO generic field (the earlier port over-read it).
 *  - dex is loaded with InMemoryDexClassLoader from BYTES read via Files.readAllBytes — sidesteps the
 *    Android 11+ "writable dex" (W^X) block that refuses to execute dex from an app-writable file. The
 *    dex is read from /sdcard/zbyd (mlstrustedobject; cameraautostudy holds MANAGE_EXTERNAL_STORAGE).
 *  - parent = the app classloader (ActivityThread.currentApplication().getClassLoader()).
 */
public final class JdwpInject {

    private static final int TAG_OBJECT = 76; // 'L'
    private static final int TAG_ARRAY  = 91; // '['

    private final AdbClient.AdbStream s;
    private int osz = 8, msz = 8, tsz = 8;
    private int id = 1;
    long tid = 0;

    public JdwpInject(AdbClient.AdbStream stream) { s = stream; }

    /** Full inject: handshake -> idsizes -> breakpoint -> wait -> read dex bytes -> InMemoryDexClassLoader -> start(). */
    public String inject(String dexPath, String unusedOptDir, String entryClass) throws Exception {
        s.write("JDWP-Handshake".getBytes("UTF-8"));
        if (!new String(s.readN(14), "UTF-8").equals("JDWP-Handshake")) throw new Exception("handshake fail");
        idsizes();

        long handlerCid = classBySig("Landroid/os/Handler;");
        long dmid = methodId(handlerCid, "dispatchMessage", "(Landroid/os/Message;)V");
        int rid = setBreakpoint((byte) 1, handlerCid, dmid);   // tag CLASS=1
        resume();
        tid = waitBreakpoint();                                // blocks until the poke triggers dispatchMessage
        clearBreakpoint(rid);

        // parent = app classloader
        long atCid = classBySig("Landroid/app/ActivityThread;");
        long camMid = methodId(atCid, "currentApplication", "()Landroid/app/Application;");
        long app = invokeStatic(atCid, camMid, new int[0], new long[0]);
        long ctxCid = classBySig("Landroid/content/Context;");
        long gclMid = methodId(ctxCid, "getClassLoader", "()Ljava/lang/ClassLoader;");
        long loader = invokeInstance(app, ctxCid, gclMid, new int[0], new long[0]);

        // read dex bytes: Files.readAllBytes(Paths.get(dexPath))  (data read; W^X applies only to executing files)
        long sp = createString(dexPath);
        long strArrTid = classBySig("[Ljava/lang/String;");
        long emptyArr = arrayNew(strArrTid, 0);
        long pathsCid = classBySig("Ljava/nio/file/Paths;");
        long getMid = methodId(pathsCid, "get", "(Ljava/lang/String;[Ljava/lang/String;)Ljava/nio/file/Path;");
        long path = invokeStatic(pathsCid, getMid, new int[]{TAG_OBJECT, TAG_ARRAY}, new long[]{sp, emptyArr});
        long filesCid = classBySig("Ljava/nio/file/Files;");
        long rabMid = methodId(filesCid, "readAllBytes", "(Ljava/nio/file/Path;)[B");
        long dexBytes = invokeStatic(filesCid, rabMid, new int[]{TAG_OBJECT}, new long[]{path});
        if (dexBytes == 0) throw new Exception("readAllBytes failed (dex unreadable: " + dexPath + ")");

        // ByteBuffer.wrap(bytes) -> new InMemoryDexClassLoader(bb, appLoader)
        long bbCid = classBySig("Ljava/nio/ByteBuffer;");
        long wrapMid = methodId(bbCid, "wrap", "([B)Ljava/nio/ByteBuffer;");
        long bb = invokeStatic(bbCid, wrapMid, new int[]{TAG_ARRAY}, new long[]{dexBytes});
        long imCid = classBySig("Ldalvik/system/InMemoryDexClassLoader;");
        long ctor = methodId(imCid, "<init>", "(Ljava/nio/ByteBuffer;Ljava/lang/ClassLoader;)V");
        long dcl = newInstance(imCid, ctor, new int[]{TAG_OBJECT, TAG_OBJECT}, new long[]{bb, loader});

        // loadClass(entry) -> reflectedType -> static start()
        long clCid = classBySig("Ljava/lang/ClassLoader;");
        long lcMid = methodId(clCid, "loadClass", "(Ljava/lang/String;)Ljava/lang/Class;");
        long klass = invokeInstance(dcl, clCid, lcMid, new int[]{TAG_OBJECT}, new long[]{createString(entryClass)});
        if (klass == 0) throw new Exception("loadClass returned null (dex not loaded)");
        long etid = reflectedType(klass);
        long start = methodId(etid, "start", null);
        invokeStatic(etid, start, new int[0], new long[0]);
        resume();
        return "ok tid=" + tid;
    }

    // ---- JDWP packet plumbing ----------------------------------------------------------------
    private byte[] cmd(int cs, int c, byte[] data) throws Exception {
        int pid = id++;
        ByteBuffer h = ByteBuffer.allocate(11 + (data == null ? 0 : data.length));
        h.putInt(11 + (data == null ? 0 : data.length)); h.putInt(pid); h.put((byte) 0); h.put((byte) cs); h.put((byte) c);
        if (data != null) h.put(data);
        s.write(h.array());
        while (true) {
            byte[] head = s.readN(11);
            ByteBuffer hb = ByteBuffer.wrap(head);
            int len = hb.getInt(); int rpid = hb.getInt(); int flags = hb.get() & 0xff; int err = hb.getShort() & 0xffff;
            byte[] body = len > 11 ? s.readN(len - 11) : new byte[0];
            if ((flags & 0x80) != 0 && rpid == pid) { if (err != 0) throw new Exception("JDWP err " + err); return body; }
            // else: async composite event (e.g. our breakpoint) — ignore while awaiting a reply
        }
    }

    private void idsizes() throws Exception {
        byte[] d = cmd(1, 7, null); ByteBuffer b = ByteBuffer.wrap(d);
        b.getInt(); msz = b.getInt(); osz = b.getInt(); tsz = b.getInt();
    }
    private long rid(ByteBuffer b) { return osz == 8 ? b.getLong() : (b.getInt() & 0xffffffffL); }
    private long rmid(ByteBuffer b) { return msz == 8 ? b.getLong() : (b.getInt() & 0xffffffffL); }
    private byte[] oid(long v) { ByteBuffer b = ByteBuffer.allocate(osz); if (osz == 8) b.putLong(v); else b.putInt((int) v); return b.array(); }
    private byte[] tidb(long v) { ByteBuffer b = ByteBuffer.allocate(tsz); if (tsz == 8) b.putLong(v); else b.putInt((int) v); return b.array(); }
    private byte[] midb(long v) { ByteBuffer b = ByteBuffer.allocate(msz); if (msz == 8) b.putLong(v); else b.putInt((int) v); return b.array(); }

    /** VirtualMachine.ClassesBySignature (1,2) -> first matching reference type id. */
    private long classBySig(String sig) throws Exception {
        byte[] sb = sig.getBytes("UTF-8");
        ByteBuffer d = ByteBuffer.allocate(4 + sb.length); d.putInt(sb.length); d.put(sb);
        byte[] r = cmd(1, 2, d.array()); ByteBuffer b = ByteBuffer.wrap(r);
        int n = b.getInt(); if (n == 0) throw new Exception("class not loaded: " + sig);
        b.get(); /* tag */ return rid(b);
    }

    /** ReferenceType.Methods (2,5): per method = methodId, name, signature, modBits (NO generic). */
    private long methodId(long refId, String name, String sigWant) throws Exception {
        byte[] r = cmd(2, 5, oid(refId)); ByteBuffer b = ByteBuffer.wrap(r);
        int n = b.getInt(); long found = 0;
        for (int i = 0; i < n; i++) {
            long mid = rmid(b);
            String nm = readStr(b), sg = readStr(b); b.getInt(); // modBits
            if (nm.equals(name) && (sigWant == null ? sg.startsWith("()") : sg.equals(sigWant))) found = mid;
        }
        if (found == 0) throw new Exception("method " + name + " not found on " + refId);
        return found;
    }
    private String readStr(ByteBuffer b) { int n = b.getInt(); byte[] s2 = new byte[n]; b.get(s2); try { return new String(s2, "UTF-8"); } catch (Exception e) { return ""; } }

    private int setBreakpoint(byte tag, long cid, long mid) throws Exception {
        ByteBuffer loc = ByteBuffer.allocate(1 + osz + msz + 8);
        loc.put(tag); loc.put(oid(cid)); loc.put(midb(mid)); loc.putLong(0);
        ByteBuffer d = ByteBuffer.allocate(6 + loc.capacity());
        d.put((byte) 2).put((byte) 2).putInt(1).put((byte) 7).put(loc.array());   // BREAKPOINT, SUSPEND_ALL, 1 mod, LocationOnly
        return ByteBuffer.wrap(cmd(15, 1, d.array())).getInt();
    }
    private void clearBreakpoint(int reqId) throws Exception { ByteBuffer d = ByteBuffer.allocate(5); d.put((byte) 2).putInt(reqId); cmd(15, 2, d.array()); }
    private void resume() throws Exception { cmd(1, 9, null); }

    /** Read events (after resume) until a BREAKPOINT composite arrives; return its thread id. */
    private long waitBreakpoint() throws Exception {
        while (true) {
            byte[] head = s.readN(11); ByteBuffer hb = ByteBuffer.wrap(head);
            int len = hb.getInt(); hb.getInt(); int flags = hb.get() & 0xff; int a = hb.get() & 0xff, c = hb.get() & 0xff;
            byte[] body = len > 11 ? s.readN(len - 11) : new byte[0];
            if ((flags & 0x80) == 0 && a == 64 && c == 100) {            // Composite event
                ByteBuffer b = ByteBuffer.wrap(body); b.get(); b.getInt(); // suspendPolicy, events count
                int kind = b.get() & 0xff; b.getInt();                    // eventKind, requestID
                long t = rid(b);
                if (kind == 2) return t;                                  // BREAKPOINT
            }
        }
    }

    private long createString(String str) throws Exception {
        byte[] sb = str.getBytes("UTF-8"); ByteBuffer d = ByteBuffer.allocate(4 + sb.length); d.putInt(sb.length); d.put(sb);
        return rid(ByteBuffer.wrap(cmd(1, 11, d.array())));
    }
    /** ArrayType.NewInstance (4,1) -> new array object id. */
    private long arrayNew(long arrTypeId, int length) throws Exception {
        ByteBuffer d = ByteBuffer.allocate(osz + 4); d.put(oid(arrTypeId)).putInt(length);
        byte[] r = cmd(4, 1, d.array()); ByteBuffer b = ByteBuffer.wrap(r); b.get(); return rid(b);
    }
    private long reflectedType(long classObj) throws Exception {
        ByteBuffer b = ByteBuffer.wrap(cmd(17, 1, oid(classObj))); b.get(); return rid(b);
    }

    /** Encode an argument vector: count + per arg (tag byte + osz-wide value). */
    private byte[] packArgs(int[] tags, long[] vals) {
        ByteBuffer a = ByteBuffer.allocate(4 + tags.length * (1 + osz));
        a.putInt(tags.length);
        for (int i = 0; i < tags.length; i++) { a.put((byte) tags[i]); a.put(oid(vals[i])); }
        return a.array();
    }

    /** ClassType.InvokeMethod (3,3) static. */
    private long invokeStatic(long typeId, long methodId, int[] tags, long[] vals) throws Exception {
        byte[] args = packArgs(tags, vals);
        ByteBuffer d = ByteBuffer.allocate(osz + tsz + msz + args.length + 4);
        d.put(oid(typeId)).put(tidb(tid)).put(midb(methodId)).put(args).putInt(1);   // INVOKE_SINGLE_THREADED
        ByteBuffer b = ByteBuffer.wrap(cmd(3, 3, d.array())); b.get(); return rid(b);
    }
    /** ObjectReference.InvokeMethod (9,6) virtual. */
    private long invokeInstance(long obj, long classId, long methodId, int[] tags, long[] vals) throws Exception {
        byte[] args = packArgs(tags, vals);
        ByteBuffer d = ByteBuffer.allocate(osz + tsz + osz + msz + args.length + 4);
        d.put(oid(obj)).put(tidb(tid)).put(oid(classId)).put(midb(methodId)).put(args).putInt(1);
        ByteBuffer b = ByteBuffer.wrap(cmd(9, 6, d.array())); b.get(); return rid(b);
    }
    /** ClassType.NewInstance (3,4). */
    private long newInstance(long classId, long ctorId, int[] tags, long[] vals) throws Exception {
        byte[] args = packArgs(tags, vals);
        ByteBuffer d = ByteBuffer.allocate(osz + tsz + msz + args.length + 4);
        d.put(oid(classId)).put(tidb(tid)).put(midb(ctorId)).put(args).putInt(1);
        ByteBuffer b = ByteBuffer.wrap(cmd(3, 4, d.array())); b.get(); return rid(b);   // tagged object value
    }
}

File: scripts/jdwp_inject.py
#!/usr/bin/env python3
# JDWP in-process dex injector (non-root).
# Loads an agent dex into a DEBUGGABLE target app and invokes <entryClass>.<entryMethod>().
# Used to run HudPrivAgent inside com.byd.cameraautostudy (holds BYDAUTO_INSTRUMENT_SET +
# BYDAUTO_SETTING_SET) so privileged BYD HAL calls pass the per-UID permission check.
#
# Mechanism: handshake -> IDSizes -> set BREAKPOINT on android.os.Handler.dispatchMessage to obtain
# a thread suspended at a Java safe-point -> CreateString + ClassLoader.getSystemClassLoader() ->
# new DexClassLoader(dexPath,optDir,null,sysLoader) -> loadClass(entry) -> ReflectedType ->
# ClassType.InvokeMethod start(). The app is poked (am start) to guarantee dispatchMessage traffic.
#
# Usage: python3 jdwp_inject.py <serial> <package> <devDexPath> [entryClass] [entryMethod]
import socket, struct, subprocess, sys, time

SERIAL   = sys.argv[1]
PKG      = sys.argv[2]
DEX      = sys.argv[3]                                   # path ON DEVICE (world-readable, e.g. /data/local/tmp/HudPrivAgent.dex)
ENTRYCLS = sys.argv[4] if len(sys.argv) > 4 else "com.zbyd.hudpriv.HudPrivAgent"
ENTRYMTD = sys.argv[5] if len(sys.argv) > 5 else "start"
OPTDIR   = "/data/local/tmp/zbyd-odex"
PORT     = 7720
POKE     = sys.argv[6] if len(sys.argv) > 6 else (PKG + "/.CameraAutoStudyTest" if PKG == "com.byd.cameraautostudy" else None)

def adb(*a, **k):
    return subprocess.run(["adb", "-s", SERIAL, *a], capture_output=True, text=True, **k)

def pid_of():
    r = adb("shell", "pidof", PKG)
    return r.stdout.strip().split()[0] if r.stdout.strip() else None

# ---- JDWP packet plumbing ----------------------------------------------------
class JDWP:
    def __init__(self, sock):
        self.s = sock; self.id = 1
        self.s.sendall(b"JDWP-Handshake")
        if self.s.recv(14) != b"JDWP-Handshake": raise RuntimeError("handshake failed")
        self.osz = self.msz = self.fsz = self.tsz = self.ssz = 8   # filled by idsizes()
        self.tid = 0

    def cmd(self, cs, c, data=b""):
        pid = self.id; self.id += 1
        pkt = struct.pack(">IIBBB", 11 + len(data), pid, 0, cs, c) + data
        self.s.sendall(pkt)
        return self.reply(pid)

    def reply(self, want):
        while True:
            head = self._recvn(11)
            length, pid, flags, a, b = struct.unpack(">IIBBB", head)
            body = self._recvn(length - 11)
            if flags & 0x80 and pid == want:
                err = (a << 8) | b
                if err: raise RuntimeError("JDWP error %d (cmd id %d)" % (err, pid))
                return body
            # else: event/composite or other reply — ignore here (we poll events separately)

    def _recvn(self, n):
        buf = b""
        while len(buf) < n:
            chunk = self.s.recv(n - len(buf))
            if not chunk: raise RuntimeError("socket closed")
            buf += chunk
        return buf

    def idsizes(self):
        d = self.cmd(1, 7)
        self.fsz, self.msz, self.osz, self.tsz, self.ssz = struct.unpack(">IIIII", d[:20])

    def oid(self, v): return v.to_bytes(self.osz, "big")
    def rd_oid(self, d, o): return int.from_bytes(d[o:o+self.osz], "big"), o + self.osz
    def rd_mid(self, d, o): return int.from_bytes(d[o:o+self.msz], "big"), o + self.msz

    def create_string(self, s):
        b = s.encode("utf-8")
        d = self.cmd(1, 11, struct.pack(">I", len(b)) + b)
        return int.from_bytes(d[:self.osz], "big")

    def classes_by_sig(self, sig):
        b = sig.encode("utf-8")
        d = self.cmd(1, 2, struct.pack(">I", len(b)) + b)
        n = struct.unpack(">I", d[:4])[0]
        if n == 0: raise RuntimeError("class not loaded: " + sig)
        # entry: refTypeTag(1) + referenceTypeID(osz) + status(4)
        tag = d[4]; cid = int.from_bytes(d[5:5+self.osz], "big")
        return tag, cid

    def methods(self, cid):
        d = self.cmd(2, 5, self.oid(cid)); o = 0
        n = struct.unpack(">I", d[o:o+4])[0]; o += 4
        out = {}
        for _ in range(n):
            mid, o = self.rd_mid(d, o)
            nl = struct.unpack(">I", d[o:o+4])[0]; o += 4; name = d[o:o+nl].decode(); o += nl
            sl = struct.unpack(">I", d[o:o+4])[0]; o += 4; sig = d[o:o+sl].decode(); o += sl
            gl = struct.unpack(">I", d[o:o+4])[0]; o += 4; o += gl
            mod = struct.unpack(">I", d[o:o+4])[0]; o += 4
            out.setdefault(name, []).append((mid, sig))
        return out

    def set_breakpoint(self, tag, cid, mid):
        # EventRequest.Set: eventKind=BREAKPOINT(2), suspend=ALL(2), 1 modifier LocationOnly(7)
        loc = struct.pack(">B", tag) + self.oid(cid) + mid.to_bytes(self.msz, "big") + struct.pack(">Q", 0)
        data = struct.pack(">BBI", 2, 2, 1) + struct.pack(">B", 7) + loc
        d = self.cmd(15, 1, data)
        return struct.unpack(">I", d[:4])[0]

    def clear_breakpoint(self, rid):
        self.cmd(15, 2, struct.pack(">BI", 2, rid))

    def resume(self):  self.cmd(1, 9)
    def suspend(self): self.cmd(1, 8)

    def wait_breakpoint(self, rid, timeout=30):
        self.s.settimeout(timeout)
        while True:
            head = self._recvn(11)
            length, pid, flags, a, b = struct.unpack(">IIBBB", head)
            body = self._recvn(length - 11)
            if not (flags & 0x80) and a == 64 and b == 100:    # Composite event
                # suspendPolicy(1) events(4) then events; first event: eventKind(1) requestID(4) thread(osz) location
                o = 1; nev = struct.unpack(">I", body[o:o+4])[0]; o += 4
                kind = body[o]; o += 1
                req = struct.unpack(">I", body[o:o+4])[0]; o += 4
                tid = int.from_bytes(body[o:o+self.osz], "big"); o += self.osz
                if kind == 2:                                  # BREAKPOINT
                    return tid

    def system_classloader(self):
        tag, cid = self.classes_by_sig("Ljava/lang/ClassLoader;")
        m = self.methods(cid)
        mid = next(mid for (mid, sig) in m["getSystemClassLoader"] if sig == "()Ljava/lang/ClassLoader;")
        # ClassType.InvokeMethod static, 0 args
        d = self.invoke_static(cid, self.tid, mid, [])
        return d  # objectID of the loader (a tagged-value 'L')

    def invoke_static(self, cid, tid, mid, args):
        data = self.oid(cid) + tid.to_bytes(self.tsz, "big") + mid.to_bytes(self.msz, "big")
        data += struct.pack(">I", len(args))
        for t, v in args: data += struct.pack(">B", t) + self.oid(v)
        data += struct.pack(">I", 1)                           # INVOKE_SINGLE_THREADED
        d = self.cmd(3, 3, data)
        # returnValue: tagged value, then exception objectID
        tagv = d[0]; val = int.from_bytes(d[1:1+self.osz], "big")
        return val

    def invoke_instance(self, obj, cid, tid, mid, args):
        data = self.oid(obj) + tid.to_bytes(self.tsz, "big") + self.oid(cid) + mid.to_bytes(self.msz, "big")
        data += struct.pack(">I", len(args))
        for t, v in args: data += struct.pack(">B", t) + self.oid(v)
        data += struct.pack(">I", 1)
        d = self.cmd(9, 6, data)
        val = int.from_bytes(d[1:1+self.osz], "big")
        return val

    def new_instance(self, cid, tid, mid, args):
        data = self.oid(cid) + tid.to_bytes(self.tsz, "big") + mid.to_bytes(self.msz, "big")
        data += struct.pack(">I", len(args))
        for t, v in args: data += struct.pack(">B", t) + self.oid(v)
        data += struct.pack(">I", 1)
        d = self.cmd(3, 4, data)
        # returns tagged object value + exception
        val = int.from_bytes(d[1:1+self.osz], "big")
        return val

    def reflected_type(self, classobj):
        d = self.cmd(17, 1, self.oid(classobj))
        tag = d[0]; tid = int.from_bytes(d[1:1+self.osz], "big")
        return tag, tid

TAG_OBJECT = 76  # 'L'

def main():
    # 1. ensure app running, get pid (poke to generate looper traffic)
    if POKE: adb("shell", "am", "start", "-n", POKE)                   # poke persistent app to run looper
    pid = pid_of()
    if not pid:
        adb("shell", "monkey", "-p", PKG, "1")
        time.sleep(1.0); pid = pid_of()
    if not pid:
        print("[x] %s not running and could not be started" % PKG); sys.exit(1)
    print("[*] %s pid=%s" % (PKG, pid))
    adb("shell", "mkdir", "-p", OPTDIR)
    adb("forward", "tcp:%d" % PORT, "jdwp:%s" % pid)
    s = socket.create_connection(("127.0.0.1", PORT), timeout=15)
    j = JDWP(s); j.idsizes()
    print("[*] connected, idsizes osz=%d msz=%d" % (j.osz, j.msz))

    # 2. breakpoint on Handler.dispatchMessage to get a suspended thread at a safe point
    tag, hcid = j.classes_by_sig("Landroid/os/Handler;")
    hm = j.methods(hcid)
    dmid = next(mid for (mid, sig) in hm["dispatchMessage"] if sig == "(Landroid/os/Message;)V")
    rid = j.set_breakpoint(tag, hcid, dmid)
    j.resume()
    print("[*] breakpoint set, poking app for looper traffic ...")
    # poke repeatedly so the main thread dispatches messages
    for _ in range(8):
        adb("shell", "input", "keyevent", "KEYCODE_WAKEUP")
        if POKE: adb("shell", "am", "start", "-n", POKE)
        try:
            j.tid = j.wait_breakpoint(rid, timeout=4)
            break
        except socket.timeout:
            continue
    else:
        print("[x] breakpoint never hit (app idle?). Try interacting with it on screen."); sys.exit(2)
    print("[*] thread suspended tid=%d" % j.tid)
    j.clear_breakpoint(rid)

    # 3. build args: dexPath, optDir, parent loader
    sp = j.create_string(DEX); op = j.create_string(OPTDIR)
    loader = j.system_classloader()
    print("[*] sysClassLoader=%d" % loader)

    # 4. new DexClassLoader(String,String,String,ClassLoader) -- pass null 3rd (libPath)
    _, dclcid = j.classes_by_sig("Ldalvik/system/DexClassLoader;")
    dclm = j.methods(dclcid)
    ctor = next(mid for (mid, sig) in dclm["<init>"]
                if sig == "(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/ClassLoader;)V")
    dcl = j.new_instance(dclcid, j.tid, ctor,
                         [(TAG_OBJECT, sp), (TAG_OBJECT, op), (TAG_OBJECT, 0), (TAG_OBJECT, loader)])
    print("[*] DexClassLoader=%d" % dcl)

    # 5. loadClass(entry) -> Class object
    clsname = j.create_string(ENTRYCLS)
    _, clcid = j.classes_by_sig("Ljava/lang/ClassLoader;")
    clm = j.methods(clcid)
    lc = next(mid for (mid, sig) in clm["loadClass"] if sig == "(Ljava/lang/String;)Ljava/lang/Class;")
    klass = j.invoke_instance(dcl, clcid, j.tid, lc, [(TAG_OBJECT, clsname)])
    print("[*] loaded %s -> classObj=%d" % (ENTRYCLS, klass))

    # 6. ReflectedType(classObj) -> typeID; find entry method; ClassType.InvokeMethod
    _, etid = j.reflected_type(klass)
    em = j.methods(etid)
    start = next(mid for (mid, sig) in em[ENTRYMTD] if sig.startswith("()"))
    ret = j.invoke_static(etid, j.tid, start, [])
    print("[*] %s.%s() invoked (ret obj=%d)" % (ENTRYCLS, ENTRYMTD, ret))

    j.resume()
    adb("forward", "--remove", "tcp:%d" % PORT)
    print("[ok] agent injected. socket @zbyd_hud_priv should now be listening.")

if __name__ == "__main__":
    main()
